### ISSUE TITLE
DokanMain CreatFile deadlock in Dokan1.dll

### ISSUE THREAD

**Issue by FaerieLee** [dokan-dev/dokany#930]
Feature request can skip this form. Bug report **must** complete it. `Check List` must be 100% match or it will be automatically closed without further discussion. Please remove this line.

# Environment
 * Windows version: 
 * Processor architecture: 
 * Dokany version: 
 * Library type (Dokany/FUSE): 
 
# Check List
  - [ ] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [ ] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [ ] I can always reproduce the issue with the provided description below.
  - [ ] I have updated Dokany to the [latest version](https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.
  - [ ] I tested one of the last [snapshot](https://github.com/dokan-dev/dokany/wiki/Build#user-snapshot) from appveyor CI

# Description
In the case of existing a document security system in my computer, I use mirror.exe, it lead to that my computer cann't shutdown. I analyse the log and find that windows kernel would deadlock when calling CreateFile to create volume in DokanMain of Doakn1.dll, the function would lock and no return. So I need help and find the deadlock reason. Thanks 

# Logs
Please attach in separate files: mirror [output](https://github.com/dokan-dev/dokany/wiki/How-to-Debug-Dokan#mirrorfs), library logs and kernel [logs](https://github.com/dokan-dev/dokany/wiki/How-to-Debug-Dokan#get-logs).
In case of BSOD, please attach minidump or dump [analyze output](https://github.com/dokan-dev/dokany/wiki/How-to-Debug-Dokan#crash-report-bsod).

**Comment by Liryna** [comment]
Hi @FaerieLee ,

Could you provide the call stack where it deadlock ?
Do you have the possibility to use windbg ? It might give you a deeper call stack when all symbols are loaded and therefore shows where is the deadlock.

**Comment by FaerieLee** [comment]
Yeah

**Comment by Liryna** [comment]
@FaerieLee I see that you are using one thread. Does it also reproduce with more threads?

**Comment by FaerieLee** [comment]
![截图](https://user-images.githubusercontent.com/18029920/94657506-2478a080-0334-11eb-8974-7c738252471e.png)

**Comment by Liryna** [comment]
Seems like it is this CreateFile that deadlock https://github.com/dokan-dev/dokany/blob/master/dokan/dokan.c#L286-L287
Before actually doing the CreateFile, Windows tries to query the volume information and this is where it deadlocks.
Which should call https://github.com/dokan-dev/dokany/blob/f7161d7f760f8bf3597c0597066192363b4d470c/sys/dispatch.c#L179
https://github.com/dokan-dev/dokany/blob/f7161d7f760f8bf3597c0597066192363b4d470c/sys/volume.c#L64

Since we do not have started thread to read the request in the library, we are going to send "default" informations. Like here for FileFsVolumeInformation https://github.com/dokan-dev/dokany/blob/f7161d7f760f8bf3597c0597066192363b4d470c/sys/volume.c#L71-L105
Would you be able to enable logs or even add more logs to see if thats actually what is happening ?

**Comment by FaerieLee** [comment]
I don't know wheher or not  it reproduce with more threads. But when i use four thread, the situation is same with one.

**Comment by Liryna** [comment]
@FaerieLee Yes looking at you callstack, we are even not at the moment the number of threads have started so thats not relevant

**Comment by FaerieLee** [comment]
Enable logs means that open dokan1.sys log ? How to open dokan1.sys log ? if I add more logs in dokan1.sys, can dokan1.sy run lacking of Authentication of Microsoft ?

**Comment by Liryna** [comment]
Here is how https://github.com/dokan-dev/dokany/wiki/How-to-Debug-Dokan#enable-logs

**Comment by FaerieLee** [comment]
![1](https://user-images.githubusercontent.com/18029920/95581590-d5391b00-0a6b-11eb-81f3-81146ecd1b1f.png)
![2](https://user-images.githubusercontent.com/18029920/95581604-da966580-0a6b-11eb-9402-49afb43723b7.png)
![3](https://user-images.githubusercontent.com/18029920/95581618-dec28300-0a6b-11eb-969b-6b4704320bdb.png)
![4](https://user-images.githubusercontent.com/18029920/95581621-e1bd7380-0a6b-11eb-9bd7-6799578ad294.png)

**Comment by FaerieLee** [comment]
Because my computer would crash when using "save"  in DbgView.exe, So I take screenshots of dokan log.

**Comment by Liryna** [comment]
Ok so it seems like the device was mounted but dismount because no thread are currently started on the library side.
You should see in the windows Event Viewer this log https://github.com/dokan-dev/dokany/blob/master/sys/event.c#L592
do you also see this one ? https://github.com/dokan-dev/dokany/blob/master/sys/event.c#L853

If not, it means something is locking the thread in this function https://github.com/dokan-dev/dokany/blob/master/sys/event.c#L566
Could you find out by adding more logs ? (or tracing with windbg if you are used to)

**Comment by Liryna** [comment]
Have you been able to get more information on the situation?

**Comment by infeo** [comment]
I'd like to add that in the Cryptomator project we are facing the same issue. There are several user reports about this deadlock (mainly during startup of the whole system, ). A simple unmount and mount does the trick, but still, but definitively an annoyance.

Is there anything i can help to debug? Unfortunately, up to now I was unable to reproduce the issue on my test setup.

**Comment by Liryna** [comment]
@infeo We need the logs from my previous message and also try to find what all those users have in common (apps / antivirus / ...) that create the issue.

**Comment by infeo** [comment]
About the above mentioned events, do you mean "Entered event start" and "Finished event start successfully"? The linked lines are a little off, so i just want to be sure.

**Comment by Liryna** [comment]
@infeo Oh indeed, yes you are right for the lines.

**Comment by infeo** [comment]
Since i still cannot reproduce it, I asked affected users.

I have two event logs, [this](https://github.com/cryptomator/cryptomator/files/5751268/Dokan1.zip) and [that](https://github.com/cryptomator/cryptomator/files/5717574/eventlogs.zip) one. They both look pretty much the same, both finished the start event, then several notifications "Handle created before IOCTL_EVENT_WAIT for file" and then the release event is triggered.

**Comment by Liryna** [comment]
Exact, without the keepalive, the device will be unmount.
I wonder if just doing the change proposed here https://github.com/dokan-dev/dokany/issues/953 would be enough.

Removing the write permission from the device createfiles and fully remove the keepalive create permission (set to 0) would be enough.
This can be tested just by changing the dokan library without building the driver.

**Comment by Liryna** [comment]
@infeo should this issue now be closed ?

**Comment by infeo** [comment]
What about the deadlock in the mirror sample application? In #953 the direct unmount was fixed, but does this also prevent the here described deadlock? @FaerieLee Can you test this with the most recent HEAD of Dokany?

**Comment by Liryna** [comment]
This should be retested with the new version https://github.com/dokan-dev/dokany/releases/tag/v1.5.0.1000 and can probably be closed.
